hazelcast/src/main/java/com/hazelcast/multimap/impl/MultiMapProxySupport.java

```diff
@@ -16,54 +16,268 @@
+import com.hazelcast.spi.impl.InternalCompletableFuture;
 import com.hazelcast.spi.impl.NodeEngine;
 import com.hazelcast.spi.impl.operationservice.Operation;
-import com.hazelcast.internal.util.ExceptionUtil;
-import com.hazelcast.internal.util.ThreadUtil;
+import com.hazelcast.spi.impl.operationservice.OperationFactory;
+import com.hazelcast.spi.impl.operationservice.OperationService;
 
+import javax.annotation.Nullable;
+import java.util.Arrays;
 import java.util.HashSet;
+import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiConsumer;
 
+import static com.hazelcast.internal.util.CollectionUtil.asIntegerList;
+import static com.hazelcast.internal.util.ConcurrencyUtil.CALLER_RUNS;
+import static com.hazelcast.internal.util.ExceptionUtil.rethrow;
 import static com.hazelcast.internal.util.MapUtil.toIntSize;
+import static com.hazelcast.internal.util.Preconditions.checkNotNull;
+import static com.hazelcast.multimap.impl.MultiMapProxyImpl.NULL_KEY_IS_NOT_ALLOWED;
+import static com.hazelcast.multimap.impl.MultiMapProxyImpl.NULL_VALUE_IS_NOT_ALLOWED;
+import static com.hazelcast.spi.impl.InternalCompletableFuture.newCompletedFuture;
+import static java.lang.Math.ceil;
+import static java.lang.Math.log10;
+import static java.util.Collections.singletonMap;
 
 public abstract class MultiMapProxySupport extends AbstractDistributedObject<MultiMapService> {
 
     protected final MultiMapConfig config;
     protected final String name;
     protected final LockProxySupport lockSupport;
+    protected final SerializationService serializationService;
+    protected final OperationService operationService;
+    protected final IPartitionService partitionService;
+    private final float putAllInitialSizeFactor;
 
     protected MultiMapProxySupport(MultiMapConfig config, MultiMapService service, NodeEngine nodeEngine, String name) {
         super(nodeEngine, service);
         this.config = config;
         this.name = name;
 
+        this.partitionService = nodeEngine.getPartitionService();
+        this.serializationService = nodeEngine.getSerializationService();
+        this.operationService = nodeEngine.getOperationService();
+
         lockSupport = new LockProxySupport(new DistributedObjectNamespace(MultiMapService.SERVICE_NAME, name),
                 LockSupportServiceImpl.getMaxLeaseTimeInMillis(nodeEngine.getProperties()));
+
+        //TODO: implement properties and properties.getFloat(MAP_PUT_ALL_INITIAL_SIZE_FACTOR);
+        this.putAllInitialSizeFactor = 1;
     }
 
     @Override
     public String getName() {
         return name;
     }
 
+    //NB: this method is copied from MapProxySupport#getPutAllInitialSize
+    @SuppressWarnings("checkstyle:magicnumber")
+    private int getPutAllInitialSize(boolean useBatching, int mapSize, int partitionCount) {
```

Comment:
this class needs some clean up to get rid of too many TODOs, eg anything related to a potential future implementation of batching can be completely removed